Allow running without sink

karabowi opened the pull request:
Replaced `LOG(fatal)` with `LOG(error)`
in `FairRootManager::RegisterImpl()`.
Now running without `sink` will not crash.
Note that in this mode also persistent branches would not be stored anywhere.

Addresses issue #1355.

---

Checklist:

* [x] Followed the [Contributing Guidelines](https://github.com/FairRootGroup/FairRoot/blob/dev/CONTRIBUTING.md)

ChristianTackeGSI commented:
Could you add a "test" by changing sampler in Tutorial3 to not use a sink? That shoukd probably fix #1355?

karabowi commented:
Added new commit with removal of sink from Tutorial3/MQ/sampler.cxx

ChristianTackeGSI commented on base/steer/FairRootManager.cxx:
```diff
@@ -173,7 +173,7 @@ void FairRootManager::RegisterImpl(const char* name, const char* folderName, T*
             fSink->RegisterImpl(name, folderName, obj);
         } else {
-            LOG(fatal) << "The sink does not exist to store persistent branches.";
+            LOG(error) << "The sink does not exist to store persistent branches.";
```
Please excuse for noticing now.

The tutorial3 "test" helped!

```
[tut3-sampler][14:46:22.509645][ERROR][FairRootManager.cxx:176:RegisterImpl] The sink does not exist to store persistent branches.
```

Please consider outputting the `name` and/or `folderName` here as well (and that toFile is `true`)? That way people might be able to not call `Register` with `toFile=true`.

rbx commented:
if it is a valid use case, then it should not be with your severity. debug or maybe warn.

ChristianTackeGSI commented:
> if it is a valid use case, then it should not be with error severity. debug or maybe warn.

:+1:

"warn" maybe. People can set `toFile=false`, if they don't have a sink. (We should take a look, where the Register in Tut3 is and fix it.)

karabowi commented:
This Register comes for sure from the EventHeader branch.

karabowi commented:
Changed the main commit message to:

Allow running without sink

Replaced `LOG(fatal)` with `LOG(debug)` in `FairRootManager::RegisterImpl()`.
Note that in this mode also persistent branches would not be stored anywhere.
Changed `FairRootManager::InitSink()` to print a `warning` and `return false`.

karabowi commented:
I still believe a warning should be printed, as in most cases running without a `sink` is user's mistake rather than user's chosen option.

ChristianTackeGSI commented:
> I still believe a warning should be printed, as in most cases running without a `sink` is user's mistake rather than user's chosen option.

Yes, warning, please. We should warn users, if they do something non-sensible.

And in relevant `Register` calls we should check for sink being set.

ChristianTackeGSI commented on base/event/FairEventHeader.cxx:
```diff
@@ -16,7 +16,8 @@
 void FairEventHeader::Register(Bool_t Persistence)
 {
-    FairRootManager::Instance()->Register("EventHeader.", "EvtHeader", this, Persistence);
+    FairRootManager::Instance()->Register(
+        "EventHeader.", "EvtHeader", this, (FairRootManager::Instance()->GetSink() ? Persistence : false));
```
I am not so happy about FairRootManager::Instance being used more. We possibly should look at the callers and implement the check there.

That said, I am okay with this for now.

karabowi replied:
So would you prefer to apply the check `GetSink` in the places where we actually `Register` the branches, i.e. in `FairRunAna*` and `FairMCApplication`?

ChristianTackeGSI replied:
Yes, I would prefer that. It would make things more straight IMHO.

And for FairMCApplication (that's FairRunSim, right?) I would then ask: Is running a simulation without a sink a supported use case? If not, then not checking the sink and letting RegisterImpl throw a warning would actually be fine (as we would warn the user about a not really supported use case).

ChristianTackeGSI commented on base/event/FairMCEventHeader.cxx:
```diff
@@ -82,7 +82,8 @@ void FairMCEventHeader::Reset()
 {
     // Dot at the end of the name is needed for splitting!!
-    FairRootManager::Instance()->Register("MCEventHeader.", "Event", this, kTRUE);
+    FairRootManager::Instance()->Register(
+        "MCEventHeader.", "Event", this, (FairRootManager::Instance()->GetSink() ? true : false));
```
clang-tidy

> redundant boolean literal in ternary expression result

```suggestion
    FairRootManager::Instance()->Register(
        "MCEventHeader.", "Event", this, FairRootManager::Instance()->GetSink());
```

ChristianTackeGSI reviewed (commented):
Please fix the clang-tidy issue.

karabowi commented:
Print `info` in the initialization, if sink not present.
Print `warning` if registering persistent branch without sink.
Register `(MC)EventHeader` only if sink present.
